File: crates/sl-frontend/src/lib/types/themes/layout.ts
import { ButtonProps } from '@/components/ui/button';
import { LucideProps } from 'lucide-react';
import * as TooltipPrimitive from '@radix-ui/react-tooltip';
import { SidebarItemProps } from '../components';
import { TooltipContentProps } from '@/components/ui/tooltip';

export interface LayoutThemeProps {
	rootContainerProps?: React.DetailedHTMLProps<
		React.HTMLAttributes<HTMLDivElement>,
		HTMLDivElement
	>;
	containerProps?: React.DetailedHTMLProps<
		React.HTMLAttributes<HTMLDivElement>,
		HTMLDivElement
	>;
	mainWindowContainerProps?: React.DetailedHTMLProps<
		React.HTMLAttributes<HTMLDivElement>,
		HTMLDivElement
	>;
	contentContainerProps?: React.DetailedHTMLProps<
		React.HTMLAttributes<HTMLDivElement>,
		HTMLDivElement
	>;
	navbarProps?: NavbarThemeProps;
	sidebarThemeProps?: SidebarThemeProps;
	profileSidebarThemeProps?: ProfileSidebarThemeProps;
}

export interface NavbarThemeProps {
	rootContainerProps?: React.DetailedHTMLProps<
		React.HTMLAttributes<HTMLElement>,
		HTMLElement
	>;
	containerProps?: React.DetailedHTMLProps<
		React.HTMLAttributes<HTMLDivElement>,
		HTMLDivElement
	>;
	minimizeButtonProps?: ButtonProps;
	maximizeButtonProps?: ButtonProps;
	closeButtonProps?: ButtonProps;
	minimizeIconProps?: React.ForwardRefExoticComponent<
		Omit<LucideProps, 'ref'> & React.RefAttributes<SVGSVGElement>
	>;
	maximizeIconProps?: React.ForwardRefExoticComponent<
		Omit<LucideProps, 'ref'> & React.RefAttributes<SVGSVGElement>
	>;
	closeIconProps?: React.ForwardRefExoticComponent<
		Omit<LucideProps, 'ref'> & React.RefAttributes<SVGSVGElement>
	>;
}

export interface SidebarItemThemeProps {
	buttonProps?: ButtonProps;
	iconProps?: React.DetailedHTMLProps<
		React.HTMLAttributes<HTMLHeadingElement>,
		HTMLHeadingElement
	>;
}

export interface SidebarContainerProps {
	tooltipProps?: React.ComponentProps<typeof TooltipPrimitive.Root>;
	tooltipTriggerProps?: React.ComponentProps<typeof TooltipPrimitive.Trigger>;
	tooltipContentProps?: TooltipContentProps;
	sidebarItemProps?: SidebarItemProps;
}

export interface SidebarThemeProps {
	rootContainerProps?: React.DetailedHTMLProps<
		React.HTMLAttributes<HTMLDivElement>,
		HTMLDivElement
	>;
	containerProps?: React.DetailedHTMLProps<
		React.HTMLAttributes<HTMLDivElement>,
		HTMLDivElement
	>;
	sidebarContainerProps?: SidebarContainerProps;
}

export interface ProfileSidebarThemeProps {}

File: crates/sl-core/src/launcher/themes/layout.rs
use serde::{Deserialize, Serialize};
use serde_json::Value;

#[derive(Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct LayoutThemeProps {
    pub root_container_props: Option<Value>,
    pub container_props: Option<Value>,
    pub main_window_container_props: Option<Value>,
    pub content_container_props: Option<Value>,
    pub navbar_props: Option<NavbarThemeProps>,
    pub sidebar_theme_props: Option<SidebarThemeProps>,
}

#[derive(Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct NavbarThemeProps {
    pub root_container_props: Option<Value>,
    pub container_props: Option<Value>,
    pub minimize_button_props: Option<Value>,
    pub maximize_button_props: Option<Value>,
    pub close_button_props: Option<Value>,
    pub minimize_icon_props: Option<Value>,
    pub maximize_icon_props: Option<Value>,
    pub close_icon_props: Option<Value>,
}

#[derive(Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SidebarThemeProps {
    pub root_container_props: Option<Value>,
    pub container_props: Option<Value>,
    pub sidebar_container_props: Option<SidebarContainerProps>,
}

#[derive(Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SidebarContainerProps {
    pub tooltip_props: Option<Value>,
    pub tooltip_trigger_props: Option<Value>,
    pub tooltip_content_props: Option<Value>,
    pub sidebar_item_props: Option<Value>,
}

#[derive(Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SidebarItemThemeProps {
    pub button_props: Option<Value>,
    pub icon_props: Option<Value>,
}